chore: improve documentation & examples (#1061)
graphql/graphiql: packages/graphiql/README.md

_/ˈɡrafək(ə)l/_ A graphical interactive in-browser GraphQL IDE. [Try the live demo](http://graphql.org/swapi-graphql).

[![](https://raw.githubusercontent.com/graphql/graphiql/HEAD/resources/graphiql.png)](http://graphql.org/swapi-graphql)

## Features

- Syntax highlighting.
- Intelligent type ahead of fields, arguments, types, and more.
- Real-time error highlighting and reporting for queries and variables.
- Automatic query and variables completion.
- Automatically adds required fields to queries.
- Documentation explorer, search, with markdown support.
- Query History using local storage
- Run and inspect query results using _any_ promise that resolves JSON results. HTTPS or WSS not required.
- Supports full [GraphQL Language Specification](https://github.com/graphql/graphql-wg): 
  - Queries, Mutations, Subscriptions, Fragments, Unions, directives, multiple operations per query, etc

## Demos

We have a few demos of `master` branch via the default netlify build (the same URL paths apply to deploy previews on PRs):

> 1. [`graphiql.min.js` demo](https://graphiql-test.netlify.com/) - the min.js bundle for the current ref

Author: these demo url paths will work after merge. use the paths with the deploy preview to see them in actuon

### GraphiQL for my GraphQL Service/HTTP Server/Etc

You may be using a runtime that already provides graphiql, or that provides it via a middleware. For example, we support [`express-graphql`](https://github.com/graphql/express-graphql)!

I would suggest a search for "graphiql <my runtime>" such as "graphiql express", "graphiql absinthe", etc to learn a potentially simpler route to setup for your environment. There are many npm packages, ruby gems, java utilities for deploying graphiql.

Here are some example searches:

- https://www.npmjs.com/search?q=graphiql - ~117 hits
- https://pypi.org/search/?q=graphiql - ~33 hits
- https://search.maven.org/search?q=graphiql - ~15 hits
- https://rubygems.org/search?utf8=%E2%9C%93&query=graphiql - ~6 hits
- https://godoc.org/?q=graphiql - ~12 hits
- https://packagist.org/?query=%22graphiql%22 - ~5 hits
- https://crates.io/search?q=graphiql - ~2 hits

> though this doesnt include runtimes or libraries where `graphiql` is used isn't referenced in the package registry search entry

Reviewer: ```suggestion
This doesn't include runtimes or libraries where GraphiQL is used but isn't referenced in the package registry search entry.
```

For an example of setting up a GraphiQL, check out the [example](../examples/graphiql-cdn/) in this repository which also includes a few useful features highlighting GraphiQL's API.

The most minimal way to set up GraphiQL is a single index.html file:

```html
<html>
  <head>
    <title>Simple GraphiQL Example</title>
    <link href="https://unpkg.com/graphiql/graphiql.min.css" rel="stylesheet" />
  </head>
  <body style="margin: 0;">
    <div id="graphiql" style="height: 100vh;"></div>

    <script crossorigin src="https://unpkg.com/react/umd/react.min.js"></script>
    <script
      crossorigin
      src="https://unpkg.com/react-dom/umd/react-dom.min.js"
    /></script>
    <script crossorigin src="https://unpkg.com/graphiql/graphiql.min.js" />

    <script>
      function graphQLFetcher(graphQLParams) {
        return fetch('https://my/graphql', {
          method: 'post',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(graphQLParams),
        }).then(function(response) {

>           return response.json();

Reviewer: ```suggestion
          return response.json().catch(function(){ return response.text()});
```

        });
      }
      ReactDOM.render(
        React.createElement(GraphiQL, { fetcher: graphQLFetcher }),
        document.getElementById('graphiql'),
      );
    </script>
  </body>
</html>
```

**Notes**:

- the inlined styles are important for ensuring GraphiQL is visible and fills the canvas.
- using `React.createElement` directly is belaborous, so follow the webpack instructions below for more highly customized implementation

### Webpack/Bundler

> **Note**: If you are having webpack issues or questions about webpack, make sure you've cross-referenced your webpack configuration with our own [webpack example](../examples/graphiql-webpack) first. We confirm that this builds on every travis CI run, and that it works end to end with every publish.

Reviewer: ```suggestion
**Note**: If you are having webpack issues or questions about webpack, make sure you've cross-referenced your webpack configuration with our own [webpack example](../examples/graphiql-webpack) first. We now have tests in CI that ensure this always builds, and we ensure it works end-to-end with every publish.
```

How do we ensure it works with every publish - manually?

Author: in CI

Reviewer: Ah; so should it be `with every merge` then, or do we have a specific CI test that only runs on publish?

Author: not sure whats best. would like shorter builds. the demo project webpack build only adds about 22s. so adding a few Cypress assertions would be easy, but for now i can just confirm by going to <deploy-preview-path.netlify>.com/webpack before merging. would be good to automate at least a basic test, but the cypress suite should be focused on the minified build

Author: currently, we just do `yarn build-demo` in CI to confirm that webpack can import and resolve the GraphiQL package, and that it can perform a webpack build without all the assets that are present in `packages/graphiql`. otherwise, build-demo will not exit 0

> Using another GraphQL service? GraphiQL is pretty easy to set up.

Reviewer: We should try and avoid things like "pretty easy" because it makes people who cannot do it due to lack of context (e.g. if they're new to Node.js) feel bad.

```suggestion
Using another GraphQL service? Here's how to get GraphiQL set up:
```

> With `yarn`:

Reviewer: ```suggestion
With `npm`:
```

Author: which reminds me, we should remove anything npm related

Author: oops, got cavalier with `sed`
